pennylane/templates/subroutines/qmc.py

```diff
@@ -0,0 +1,358 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the ``QuantumMonteCarlo`` template and utility functions.
+"""
+import numpy as np
+
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+def probs_to_unitary(probs):
+    r"""Calculates the unitary matrix corresponding to an input probability distribution.
+
+    For a given distribution :math:`p_{i}`, this function returns the unitary :math:`\mathcal{A}`
+    that transforms the :math:`|0\rangle` state as
+
+    .. math::
+
+        \mathcal{A} |0\rangle = \sum_{i} \sqrt{p_{i}} |i\rangle,
+
+    so that measuring the resulting state in the computational basis will give the state
+    :math:`|i\rangle` with probability :math:`p_{i}`. Note that the returned unitary matrix is
+    real and hence an orthogonal matrix.
+
+    Args:
+        probs (array): input probability distribution as a flat array
+
+    Returns:
+        array: unitary
+
+    Raises:
+        ValueError: if the input array is not flat or does not correspond to a probability
+            distribution
+
+    **Example:**
+
+    >>> p = np.ones(4) / 4
+    >>> probs_to_unitary(p)
+    array([[ 0.5       ,  0.5       ,  0.5       ,  0.5       ],
+           [ 0.5       , -0.83333333,  0.16666667,  0.16666667],
+           [ 0.5       ,  0.16666667, -0.83333333,  0.16666667],
+           [ 0.5       ,  0.16666667,  0.16666667, -0.83333333]])
+    """
+    if not np.allclose(sum(probs), 1) or min(probs) < 0:
+        raise ValueError(
+            "A valid probability distribution of non-negative numbers that sum to one"
+            "must be input"
+        )
+
+    dim = len(probs)
+    unitary = np.zeros((dim, dim))
+
+    unitary[:, 0] = np.sqrt(probs)
+    unitary = np.linalg.qr(unitary)[0]
```

Comment:
Clever!

I'm a bit worried that this will be unnecessarily inefficient for large unitaries. Is there a more efficient way to populate the other rows and columns without having to perform a QR decomposition? :thinking:

Reply:
Are we considering including MottonenStatePreparation here?

Reply:
I guess it is not as easy as I initially guessed to build a unitary starting from just one row :thinking: 

I found [this method](https://quantumcomputing.stackexchange.com/questions/10239/how-can-i-fill-a-unitary-knowing-only-its-first-column) after a quick search, which seems to be faster than a QR decomposition. Maybe it's overkill, but it would be nice to benchmark both methods as we scale system size.

Reply:
Awesome, yes I will try this method!